Enable Export with Excel Format

positiveVibes4all opened the pull request:
## Summary
This PR adds the capability to use an export with '.xlsx' format.

positiveVibes4all committed:
Add Excel Export Capability

positiveVibes4all committed:
Bump

positiveVibes4all committed:
Cleanup

positiveVibes4all committed:
Cleanup

stellarhoof commented on packages/react/src/FilterButtonList.js:
```diff
@@ -10,6 +10,7 @@ import styles from './styles/index.js'
 import { newNodeFromField } from './utils/search.js'
 import { unusedOptions } from './FilterAdder.js'
 import { fieldsToOptions } from './utils/fields.js'
+import { displayLabelFn } from './utils/format.js'
```
I think these `displayLabelFn` changes snuck in from another branch of yours

positiveVibes4all replied:
Possibly, I will remove it. thanks for catching this.

positiveVibes4all replied:
There is way more than just that, it all made it in, I will clean up this PR

stellarhoof commented on packages/export/src/excel.js:
```diff
@@ -0,0 +1,73 @@
+      (value) => ({ value, fontWeight: 'bold' }),
+      headers || transformLabels(transform)
+    ),
+  ]
```
Will this work if `headers` is an array of arrays?

positiveVibes4all replied:
the headers are already being used like this in the CSV file from what I remember, so I did not consider them to be an array of arrays, if that was the case the CSV would not work either from what I read about that library. Can you give an example you have in mind for me to test?

stellarhoof replied:
I don't have an example I was just going off of the comment for the `headers` argument

> array of strings to use as headers, array or arrays for multi-line headers

positiveVibes4all replied:
I talked with Kathryn if there is a use case for multi-line headers today and there is not. There could be one if we supported exports of report builder, but that is on the table for discussion. We could support it if the solution we land on winds up including this, but in general for multi header exports we would need a way to support spanning multiple columns and things like this. So for now, I think I will just change this comment to a TODO in both CSV and Excel for the future when we support it from data transfer to export.

stellarhoof commented on packages/export/src/excel.js:
```diff
@@ -0,0 +1,73 @@
+const convertToExcelCell = (value) => {
+  return {
+    type: String,
+    typeof: 'string',
```
What's this `typeof` for?

positiveVibes4all replied:
left over from something I was using for column width, but I abandoned coercing values because already had some odd instances ie: zip codes that were being coerced to strings and numbers because some have a - within them.

I will remove it

dubiousdavid replied:
Reminder that this still needs to be done.

stellarhoof commented on packages/export/src/excel.js:
Can you write tests for this file

positiveVibes4all replied:
What type of test would you suggest? the data is binary, so we really cannot build one like is done for CSV's, I thought about it but it felt like there was little value add for doing this based on this constraint.

stellarhoof replied:
You can read the stream via https://www.npmjs.com/package/read-excel-file, which is nice as you don't need to write a file in the test, keeping it pure.

positiveVibes4all replied:
This seems ok, I am not in love with it because we are using another library to test this one in essence but seems like a start.

stellarhoof replied:
It's by the same author so I think that reduces the risk somewhat

positiveVibes4all replied:
I agree, it just feels like we are testing the library itself instead of anything novel about how it is used, I will add it though as it is at least something. Thanks for the feedback on this.

stellarhoof replied:
My thinking is that there is some code that can benefit from testing which is not library related. Another approach is to take a `writeData` callback that looks like

```
      const readStream = await writeXlsxFile(excelData, { columns })
      for await (const chunk of readStream) {
        stream.write(chunk)
      }
```

That way you don't have to read the xlsx file but can still test this function.

dubiousdavid replied:
Reminder that this still needs to be done.

positiveVibes4all replied:
This is in the current PR and the tests is using it as such.

positiveVibes4all committed:
Addressed PR Feedback

stellarhoof commented on packages/export/src/excel.js:
```diff
@@ -60,9 +67,7 @@ export default ({
-      for await (const chunk of readStream) {
-        stream.write(chunk)
-      }
+      await writeStreamData(stream, readStream)
```
There's not much point to this function. My suggestion above was so that you could avoid the call to `writeXlsxFile` in this function and instead pass that + writing the stream as an argument. That way you can write a unit test that passes a dummy function that doesn't use `writeXlsxFile` and you can test the functionality without actually writing an excel file.

positiveVibes4all replied:
I updated the test and code to reflect this.

positiveVibes4all committed:
Update Test Per Feedback

dubiousdavid commented on packages/export/src/csv.js:
```diff
@@ -39,6 +40,7 @@ export default ({
         recordsWritten = recordsWritten + _.getOr(1, 'recordCount', r)
         await onWrite({ recordsWritten, record: r })
       }
+      await onWrite({ recordsWritten: recordsWritten, isStreamDone: true })
```
You can use the shorthand syntax here: 
```ts
onWrite({ recordsWritten, isStreamDone: true })
```

dubiousdavid commented on packages/export/src/excel.js:
```diff
@@ -0,0 +1,75 @@
+
+const convertToExcelCell = (value) => {
+  return {
+    type: String,
```
Is there a reason we are always setting the cell type to string? According to the documentation, the following is true:

_If a cell doesn't have a type, then it is automatically detected from the value, or defaults to a String._

positiveVibes4all replied:
there are some odd cases like zip code for instance, where some of it was treated as a number and some as a string, this changes the alignment of the values and possibly other underlying things. When I get done with the other changes I can test it again, but this seems safer atm.

positiveVibes4all replied:
I double-checked this, and it handles it fine as it relies on the application to use all the data, I must have originally tested it with some other conditions that are no longer a concern. Updated the code with this change.

dubiousdavid replied:
That's fine. We can circle back to this.

dubiousdavid commented on packages/export/src/utils.js:
```diff
@@ -1,6 +1,13 @@
 import F from 'futil'
 import _ from 'lodash/fp.js'
 
+export const writeStreamData = async (stream, writeStreamData) => {
```
I think this function belongs in /export/src/excel.js. Your test for it lives in /export/src/excel.test.js and it is not used anywhere outside of the excel.js file.

dubiousdavid commented on packages/export/src/excel.js:
```diff
@@ -0,0 +1,75 @@
+        stream,
+        async () => await writeXlsxFile(excelData, { columns })
+      )
+      await onWrite({ recordsWritten: recordsWritten, isStreamDone: true })
```
You can use the shorthand syntax here: 
```ts
onWrite({ recordsWritten, isStreamDone: true })
```

dubiousdavid commented on packages/export/package.json:
```diff
@@ -34,7 +34,9 @@
     "lodash": "^4.17.21",
-    "minimal-csv-formatter": "^1.0.15"
+    "minimal-csv-formatter": "^1.0.15",
+    "read-excel-file": "^5.8.1",
```
This package is not being used.

dubiousdavid commented on packages/export/src/excel.js:
```diff
@@ -0,0 +1,75 @@
+      }
+      let columns = _.map(
+        () => ({
+          width: 50,
```
What is the default behavior for width? If the default behavior is not smart, we should consider setting the width to the largest value of the following: the length of the column name or 50.

positiveVibes4all replied:
default is not smart, 50 seemed like a reasonable size to start with and wrapping the data set to true.

positiveVibes4all replied:
I wrote some code to keep track of the longest value and then use the column name as the shortest

positiveVibes4all committed:
PR Feedback

positiveVibes4all committed:
Fixes

positiveVibes4all committed:
Smart Column Width

- Added smart column width
- Added Sticky Header
- Added Background Color like CSV for Header

positiveVibes4all committed:
Add Index Column BG Color

positiveVibes4all committed:
Update Tests

stellarhoof commented:
Code looks good. Pending approval as I cannot test this with the application due to some bug in the spark PR.

positiveVibes4all committed:
Merge branch 'main' into feature/ADD-EXCEL-EXPORT-ABILITY
